src/blog/better-class-features.md

```diff
@@ -0,0 +1,332 @@
+
+![Access of private methods](/_img/better-class-features/access-private-methods.svg)
+
+Take this snippet as an example:
+
+```js
+class A {
+  #a() {}
+}
+```
+
+V8 used to generate the following bytecode for the constructor of `A`:
+
+```cpp
+// Load the private brand symbol for class A from the context
+// and store it into r1.
+LdaImmutableCurrentContextSlot [3]
+Star r1
+
+// Load the target into r0.
+Mov <this>, r0
+// Load the current context into r2.
+Mov <context>, r2
+// Call the runtime %AddPrivateBrand() function to store the context in
+// the instance with the private brand as key.
+CallRuntime [AddPrivateBrand], r0-r2
+```
+
+Since there was also a call to the runtime function `%AddPrivateBrand()`, the overhead made the constructor much slower than constructors of classes with only public methods.
+
+### Optimizing initialization of private brands
+
+To speed up the installation of the private brands, in most cases we just reuse the `DefineKeyedOwnProperty` machinery added for the optimization of private fields:
+
+```cpp
+// Load the private brand symbol for class A from the context
+// and store it into r1
+LdaImmutableCurrentContextSlot [3]
+Star0
+
+// Use the DefineKeyedOwnProperty bytecode to store the
+// context in the instance with the private brand as key
+Ldar <context>
+DefineKeyedOwnProperty <this>, r0, [0]
+```
+
+![Performance of initialization of instances with private methods](/_img/better-class-features/private-methods-performance.svg)
+
+There is a caveat, however: if the class is a derived class whose constructor calls `super()`,  the initialization of the private methods - and in our case, the installation of the private brand symbol - has to happen after `super()` returns:
+
+```js
+class A {
+  constructor() {
+    // This throws from a new B() call because super() has not yet returned.
+    this.callMethod();
+  }
+}
+
+class B extends A {
+  #method() {}
+  callMethod() { return this.#method(); }
+  constructor(o) {
+    super();
+  }
+};
+```
+
+As described before, when initializing the brand, V8 also stores a reference to the class context in the instance. This reference isn't used in brand checks, it's intended for the debugger to retrieve a list of private methods from the instance without knowing which class it is constructed from. When `super()` is invoked directly in the constructor, V8 can simply load the context from the context register (which is what `Mov <context>, r2` or `Ldar <context>` in the bytecodes above does) to perform the initialization, but `super()` can also be invoked from a nested arrow function, which in turn can be invoked from a different context. In this case, V8 falls back to a runtime function (still named `%AddPrivateBrand()`) to look for the class context in the context chain instead of relying on the context register. For example, for the `callSuper` function below:
```

Comment:
```suggestion
As described before, when initializing the brand, V8 also stores a reference to the class context in the instance. This reference isn't used in brand checks, but is instead intended for the debugger to retrieve a list of private methods from the instance without knowing which class it is constructed from. When `super()` is invoked directly in the constructor, V8 can simply load the context from the context register (which is what `Mov <context>, r2` or `Ldar <context>` in the bytecodes above does) to perform the initialization, but `super()` can also be invoked from a nested arrow function, which in turn can be invoked from a different context. In this case, V8 falls back to a runtime function (still named `%AddPrivateBrand()`) to look for the class context in the context chain instead of relying on the context register. For example, for the `callSuper` function below:
```